test/lit/passes/remove-unused-brs_all-features.wast

```diff
@@ -115,111 +113,4 @@
- ;; CHECK-NEXT:  (call $log
- ;; CHECK-NEXT:   (i32.const 0)
- ;; CHECK-NEXT:  )
- ;; CHECK-NEXT:  (block $null
- ;; CHECK-NEXT:   (drop
- ;; CHECK-NEXT:    (ref.func $br_on-to-br)
- ;; CHECK-NEXT:   )
- ;; CHECK-NEXT:   (call $log
- ;; CHECK-NEXT:    (i32.const 1)
- ;; CHECK-NEXT:   )
- ;; CHECK-NEXT:  )
- ;; CHECK-NEXT:  (call $log
- ;; CHECK-NEXT:   (i32.const 2)
- ;; CHECK-NEXT:  )
- ;; CHECK-NEXT:  (drop
- ;; CHECK-NEXT:   (block $func (result (ref $ref|func|_=>_none))
- ;; CHECK-NEXT:    (drop
- ;; CHECK-NEXT:     (br $func
- ;; CHECK-NEXT:      (ref.func $br_on-to-br)
- ;; CHECK-NEXT:     )
- ;; CHECK-NEXT:    )
- ;; CHECK-NEXT:    (call $log
- ;; CHECK-NEXT:     (i32.const 3)
- ;; CHECK-NEXT:    )
- ;; CHECK-NEXT:    (ref.func $br_on-to-br)
- ;; CHECK-NEXT:   )
- ;; CHECK-NEXT:  )
- ;; CHECK-NEXT:  (call $log
- ;; CHECK-NEXT:   (i32.const 4)
- ;; CHECK-NEXT:  )
- ;; CHECK-NEXT:  (drop
- ;; CHECK-NEXT:   (block $i31 (result (ref i31))
- ;; CHECK-NEXT:    (drop
- ;; CHECK-NEXT:     (br $i31
- ;; CHECK-NEXT:      (i31.new
- ;; CHECK-NEXT:       (i32.const 42)
- ;; CHECK-NEXT:      )
- ;; CHECK-NEXT:     )
- ;; CHECK-NEXT:    )
- ;; CHECK-NEXT:    (call $log
- ;; CHECK-NEXT:     (i32.const 5)
- ;; CHECK-NEXT:    )
- ;; CHECK-NEXT:    (i31.new
- ;; CHECK-NEXT:     (i32.const 1337)
- ;; CHECK-NEXT:    )
- ;; CHECK-NEXT:   )
- ;; CHECK-NEXT:  )
- ;; CHECK-NEXT:  (call $log
- ;; CHECK-NEXT:   (i32.const 6)
- ;; CHECK-NEXT:  )
- ;; CHECK-NEXT:  (drop
- ;; CHECK-NEXT:   (block $non-null (result (ref $ref|func|_=>_none))
- ;; CHECK-NEXT:    (br $non-null
- ;; CHECK-NEXT:     (ref.func $br_on-to-br)
- ;; CHECK-NEXT:    )
- ;; CHECK-NEXT:    (call $log
- ;; CHECK-NEXT:     (i32.const 7)
- ;; CHECK-NEXT:    )
- ;; CHECK-NEXT:    (ref.func $br_on-to-br)
- ;; CHECK-NEXT:   )
- ;; CHECK-NEXT:  )
- ;; CHECK-NEXT: )
- (func $br_on-to-br (param $func (ref func))
-  (call $log (i32.const 0))
-  (block $null
-   ;; a non-null reference is not null, and the br is never taken
-   (drop
-    (br_on_null $null (ref.func $br_on-to-br))
-   )
-   (call $log (i32.const 1))
-  )
-  (call $log (i32.const 2))
-  (drop
-   (block $func (result funcref)
-    ;; a non-null function reference means we always take the br
-    (drop
-     (br_on_func $func (ref.func $br_on-to-br))
-    )
-    (call $log (i32.const 3))
-    (ref.func $br_on-to-br)
-   )
-  )
-  (call $log (i32.const 4))
-  (drop
-   (block $i31 (result i31ref)
-    ;; a non-null i31 reference means we always take the br
-    (drop
-     (br_on_i31 $i31
-      (i31.new (i32.const 42))
-     )
-    )
-    (call $log (i32.const 5))
-    (i31.new (i32.const 1337))
-   )
-  )
-  (call $log (i32.const 6))
-  (drop
-   (block $non-null (result (ref func))
-    ;; a non-null reference is not null, and the br is always taken
-    (br_on_non_null $non-null (ref.func $br_on-to-br))
-    (call $log (i32.const 7))
-    (ref.func $br_on-to-br)
-   )
```

Comment:
`br_on_func|i31` are legacy and I get why they are removed, but why remove `null|non_null`?

Reply:
I've added a test for `br_on_non_null` back in as well as a new test for `br_on_null`. It turns out we don't optimize `br_on_null` yet, so I left a TODO for that.

Reply:
I think we could use some more tests here. The i31 and func tests were placeholders for general `br_on` operations. How about aside from `br_on_null|non_null` we also replace the old tests with tests of `br_on_cast` which is definitely not or definitely taken?

Reply:
(or do those exist already elsewhere?)

Reply:
These are already thoroughly tested in test/lit/passes/remove-unused-brs-gc.wast